clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -709,7 +665,49 @@ private void discard() {
         }
     }
 
-    private class FetchManagerMetrics {
+    private static class FetchCompletion {
+        private final TopicPartition partition;
+        private final long fetchedOffset;
+        private final FetchResponse.PartitionData partitionData;
+        private final FetchMetricAggregator metricAggregator;
+
+        public FetchCompletion(TopicPartition partition,
+                               long fetchedOffset,
+                               FetchResponse.PartitionData partitionData,
+                               FetchMetricAggregator metricAggregator) {
+            this.partition = partition;
+            this.fetchedOffset = fetchedOffset;
+            this.partitionData = partitionData;
+            this.metricAggregator = metricAggregator;
+        }
+    }
+
+    private static class FetchMetricAggregator {
```

Comment:
Would it be worth adding a comment explaining how this works? It relies on a number of assumptions (one should call `record` once per partition and the metrics will be recorded when `record` has been called for every partition.